engine-standalone-tracing/src/types/call_tracer.rs

```diff
@@ -0,0 +1,288 @@
+                code_address,
+                transfer,
+                input,
+                target_gas,
+                is_static,
+                context,
+            } => {
+                let call_type = if is_static {
+                    CallType::StaticCall
+                } else if transfer.is_none() {
+                    CallType::DelegateCall
+                } else if code_address == context.address {
+                    CallType::Call
+                } else {
+                    CallType::CallCode
+                };
+
+                self.enter(
+                    call_type,
+                    Address::new(context.caller),
+                    Address::new(context.address),
+                    input.to_vec(),
+                    target_gas.unwrap_or_default(),
+                    context.apparent_value,
+                );
+            }
+            evm::tracing::Event::Create {
+                caller,
+                address,
+                scheme,
+                value,
+                init_code,
+                target_gas,
+            } => {
+                let call_type = match scheme {
+                    evm::CreateScheme::Legacy { .. } => CallType::Create,
+                    evm::CreateScheme::Create2 { .. } => CallType::Create2,
+                    evm::CreateScheme::Fixed(_) => CallType::Create, // is this even possible in production?
+                };
+
+                self.enter(
+                    call_type,
+                    Address::new(caller),
+                    Address::new(address),
+                    init_code.to_vec(),
+                    target_gas.unwrap_or_default(),
+                    value,
+                );
+            }
+            evm::tracing::Event::Suicide {
+                address,
+                target,
+                balance,
+            } => {
+                // TODO: gas = 0 is right?
+                self.enter(
+                    CallType::SelfDestruct,
+                    Address::new(address),
+                    Address::new(target),
+                    Vec::new(),
+                    0,
+                    balance,
+                );
+                self.exit(Vec::new(), None);
+            }
+            // Exit event always comes after RecordRefund, so we don't need to worry about gas here (it's handled in RecordRefund)
+            evm::tracing::Event::Exit {
+                reason,
+                return_value,
+            } => {
+                let error = match reason {
+                    evm::ExitReason::Succeed(_) => None,
+                    other => Some(other),
+                };
+                self.exit(return_value.to_vec(), error);
+            }
+
+            // not useful
+            evm::tracing::Event::TransactCall { .. } => (),
+            evm::tracing::Event::TransactCreate { .. } => (),
+            evm::tracing::Event::TransactCreate2 { .. } => (),
```

Comment:
Out of curiosity, why are these not useful?

Reply:
Because the same information is always duplicated in the normal `Call`/`Create` events (above code).

Reply:
ah. I see. So, when it gets to it, it already would basically be returned.

Reply:
I don't really understand what is going on here. I did notice how `call_type` was computed in the `Call` branch, but I'm not sure why is it happening this way.

Is it maybe that StaticCall/etc... is only triggered inside the context of a Call?

Reply:
I chose to take this approach because all the information I need is present in the `Call`/`Create` events. It would be possible to do this by looking at the `Step` events in the `evm_runtime` module (presently handled as a no-op), and then the opcode would be directly available. But that approach seemed more complex to me overall because the `Step` events are so fine-grained.

The long story short here is that the SputnikVM tracing is structured differently than geth's tracing, and there are multiple ways of mapping one onto the other.